File: src/lexical_analyzer.cpp
#include "lexical_analyzer.h"

#include <cctype>
#include <string>
#include <unordered_map>

#include "data/symbol_table.h"
#include "data/token_list.h"
#include "enums/grammar_symbols.h"

const std::unordered_map<std::string, Terminal> LexicalAnalyzer::_TOKEN_STRINGS = {
    {"(", OPEN_P},       {")", CLOSE_P},      {"[", OPEN_SB},        {"]", CLOSE_SB},   {"{", OPEN_CB},
    {"}", CLOSE_CB},     {",", COMMA},        {";", SEMICOLON},      {"=", ASSIGNMENT}, {"<", LESS_THAN},
    {">", GREATER_THAN}, {"<=", LESS_EQUAL},  {">=", GREATER_EQUAL}, {"==", EQUAL},     {"!=", NOT_EQUAL},
    {"*", MULTIPLY},     {"/", DIVISION},     {"%", REMAINDER},      {"+", ADDITION},   {"-", SUBTRACT},
    {"def", DEF},        {"new", NEW},        {"int", INT},          {"float", FLOAT},  {"string", STRING},
    {"print", PRINT},    {"read", READ},      {"return", RETURN},    {"break", BREAK},  {"if", IF},
    {"else", ELSE},      {"for", FOR},        {"and", AND},          {"or", OR},        {"not", NOT},
    {"call", CALL},      {"null", NULL_CONST}};

Terminal LexicalAnalyzer::check_identifier(unsigned int current_line,
                                           std::string current_token,
                                           SymbolTable &symbol_table) {
    // Checks if token is a keyword.
    // If so, returns corresponding kw;
    // Otherwise adds it to symbol table if not already in it, returning IDENT.
    if (_TOKEN_STRINGS.find(current_token) != _TOKEN_STRINGS.end())
        return _TOKEN_STRINGS.at(current_token);
    symbol_table.add_instance(current_token, current_line);
    return IDENT;
};

LexicalReturnCode LexicalAnalyzer::analyze(std::string source, TokenList &token_list, SymbolTable &symbol_table) {
    TokenParserStates state = WHITE_SPACE;
    unsigned int current_line = 1, current_column = 1, token_column = 1, current_pos = 0, size = source.size();
    char next;
    std::string current_token;

    while (current_pos < size) {
        next = source[current_pos];
        switch (state) {
            case WHITE_SPACE:
                if (isspace(next)) {
                    if (next == '\n') {
                        current_line++;
                        current_column = 0;
                    }
                    current_pos++;
                    current_column++;
                } else {
                    state = READ_SIMPLE;
                }
                break;
            case READ_SIMPLE:
                token_column = current_column;
                if (isalpha(next) || next == '_') {  // start of ident or keyword
                    state = MAYBE_IDENT;
                    current_token += next;
                } else if (isdigit(next)) {  // start of number
                    state = NUMERAL;
                    current_token += next;
                } else if (next == '\"') {  // start of string
                    state = STRING_LITERAL;
                } else if (next == '=') {
                    state = ASSIGN_OR_EQUAL;
                } else if (next == '>') {
                    state = GREATER_OR_GE;
                } else if (next == '<') {
                    state = LESS_OR_LE;
                } else if (next == '!') {
                    state = DIFFERENT;
                    // if token is 1-char and not < | > | =
                } else {
                    std::string key = std::string(1, next);

                    if (_TOKEN_STRINGS.find(key) != _TOKEN_STRINGS.end() && _TOKEN_STRINGS.at(key) < ASSIGNMENT) {
                        Terminal current_symbol = _TOKEN_STRINGS.at(key);
                        token_list.push_back(current_symbol, current_line, token_column);
                        state = WHITE_SPACE;
                    } else {
                        return INVALID_CHAR;  // error: invalid character in line N
                    }
                }
                current_pos++;
                current_column++;
                break;
            case MAYBE_IDENT:
                if (isalnum(next) || next == '_') {
                    current_token += next;
                    current_pos++;
                    current_column++;
                } else {
                    state = WHITE_SPACE;
                    Terminal result = check_identifier(current_line, current_token, symbol_table);
                    std::string data = result == IDENT ? current_token : "";
                    token_list.push_back(result, current_line, token_column, data);
                    current_token.clear();
                }
                break;
            case NUMERAL:
                if (isdigit(next)) {
                    current_token += next;
                    current_pos++;
                    current_column++;
                } else if (next == '.') {
                    state = FLOAT_FRACTIONAL;
                    current_token += next;
                    current_pos++;
                    current_column++;
                } else if (isalpha(next) || next == '_') {
                    return INVALID_NUMBER_FORMAT;
                } else {
                    state = WHITE_SPACE;
                    token_list.push_back(INT_CONST, current_line, token_column, current_token);
                    current_token.clear();
                }
                break;
            case FLOAT_FRACTIONAL:
                if (isdigit(next)) {
                    current_token += next;
                    current_pos++;
                    current_column++;
                } else if (isalpha(next) || next == '_') {
                    return INVALID_NUMBER_FORMAT;
                } else {
                    state = WHITE_SPACE;
                    token_list.push_back(FLOAT_CONST, current_line, token_column, current_token);
                    current_token.clear();
                }
                break;
            case STRING_LITERAL:  // You can't escape! (it doesnt escape chars)
                if (next == '\"') {
                    state = WHITE_SPACE;
                    token_list.push_back(STRING_CONST, current_line, token_column, current_token);
                    current_token.clear();
                } else if (next == '\n') {  // no multiline strings
                    return UNCLOSED_STRING;
                } else {
                    current_token += next;
                }
                current_pos++;
                current_column++;
                break;
            case ASSIGN_OR_EQUAL:
                if (next == '=') {
                    token_list.push_back(EQUAL, current_line, token_column);
                    current_pos++;
                    current_column++;
                } else {
                    token_list.push_back(ASSIGNMENT, current_line, token_column);
                }
                state = WHITE_SPACE;
                break;
            case GREATER_OR_GE:
                if (next == '=') {
                    token_list.push_back(GREATER_EQUAL, current_line, token_column);
                    current_pos++;
                    current_column++;
                } else {
                    token_list.push_back(GREATER_THAN, current_line, token_column);
                }
                state = WHITE_SPACE;
                break;
            case LESS_OR_LE:
                if (next == '=') {
                    token_list.push_back(LESS_EQUAL, current_line, token_column);
                    current_pos++;
                    current_column++;
                } else {
                    token_list.push_back(LESS_THAN, current_line, token_column);
                }
                state = WHITE_SPACE;
                break;
            case DIFFERENT:
                if (next == '=') {
                    current_token += next;
                    current_pos++;
                    current_column++;
                    state = WHITE_SPACE;
                    token_list.push_back(NOT_EQUAL, current_line, token_column);
                    current_token.clear();
                } else
                    return INVALID_CHAR;  // error: invalid '!' in line N
                break;
            default:
                printf("\n\nwhat\n\n");  // GREED /// THIRD : A SHOT IN THE DARK
                break;
        }
    }

    return LEX_OK;
}

File: source/lexical_analyzer.py
"""
Analisador léxico.
"""

from enum import Enum

from data.token import Token
from data.grammar import Terminal, STRING_TO_TERMINAL, ONE_CHAR_DEFINITIVE_TERMINALS
from symbol_table import SymbolTable

from return_status import LexicalReturnStatus

class LexicalAnalyzerStates(Enum):
    """
    Estados da análise.
    """

    WHITE_SPACE = 0       # You have been living here for as long as you can remember
    READ_SIMPLE = 1       # catches 1-char tokens, sends to other states
    MAYBE_IDENT = 2       # reads a word, then checks if it's a keyword
    NUMERAL = 3           # starts in a digit, may be int or float
    FLOAT_FRACTIONAL = 4  # fractional portion of float
    STRING_LITERAL = 5    # looks for string termination
    ASSIGN_OR_EQUAL = 6   # may find = or ==
    GREATER_OR_GE = 7     # may find > or >=
    LESS_OR_LE = 8        # may find < or <=
    DIFFERENT = 9         # must find = after !

class LexicalAnalyzer:
    """
    Analisador léxico.
    """

    @staticmethod
    def analyze(source: str) -> tuple[LexicalReturnStatus, tuple[int,int], list[Token], SymbolTable]:
        """
        Analisa o código fonte e retorna a lista de tokens.
        """

        source += ' '
        symbol_table = SymbolTable()
        token_list: list[Token] = []
        state = LexicalAnalyzerStates.WHITE_SPACE
        current_line = 1
        current_column = 1
        token_column = 1
        index = 0
        current_token = ''

        while index < len(source):
            next_ = source[index]

            match (state):
                case LexicalAnalyzerStates.WHITE_SPACE:
                    if next_ in ' \n':
                        if next_ == '\n':
                            current_line += 1
                            current_column = 0

                        index += 1
                        current_column += 1
                    else:
                        state = LexicalAnalyzerStates.READ_SIMPLE
                case LexicalAnalyzerStates.READ_SIMPLE:
                    token_column = current_column

                    if next_.isalpha() or next_ == '_':
                        state = LexicalAnalyzerStates.MAYBE_IDENT
                        current_token += next_
                    elif next_.isdigit():
                        state = LexicalAnalyzerStates.NUMERAL
                        current_token += next_
                    elif next_ == '"':
                        state = LexicalAnalyzerStates.STRING_LITERAL
                    elif next_ == '=':
                        state = LexicalAnalyzerStates.ASSIGN_OR_EQUAL
                    elif next_ == '>':
                        state = LexicalAnalyzerStates.GREATER_OR_GE
                    elif next_ == '<':
                        state = LexicalAnalyzerStates.LESS_OR_LE
                    elif next_ == '!':
                        state = LexicalAnalyzerStates.DIFFERENT
                    else:
                        # TODO: Otimizar isso aqui, daria pra só comparar pra ver se o next é um dos ONE_CHAR_DEFINITIVE_TERMINALS
                        if next_ in STRING_TO_TERMINAL and STRING_TO_TERMINAL[next_] in ONE_CHAR_DEFINITIVE_TERMINALS:
                            state = LexicalAnalyzerStates.WHITE_SPACE
                            token_list.append(Token(STRING_TO_TERMINAL[next_], current_line, token_column))
                        else:
                            return LexicalReturnStatus.INVALID_CHAR, (current_line,current_column), token_list, symbol_table

                    index += 1
                    current_column += 1
                case LexicalAnalyzerStates.MAYBE_IDENT:
                    if next_.isalnum() or next_ == '_':
                        current_token += next_
                        index += 1
                        current_column += 1
                    else:
                        state = LexicalAnalyzerStates.WHITE_SPACE

                        if current_token in STRING_TO_TERMINAL:
                            token_list.append(Token(STRING_TO_TERMINAL[current_token], current_line, token_column))
                        else:
                            token_list.append(Token(Terminal.IDENT, current_line, token_column, current_token))
                            symbol_table.add_symbol_instance(current_token, current_line, token_column)
                        current_token = ''
                case LexicalAnalyzerStates.NUMERAL:
                    if next_.isdigit():
                        current_token += next_
                        index += 1
                        current_column += 1
                    elif next_ == '.':
                        state = LexicalAnalyzerStates.FLOAT_FRACTIONAL
                        current_token += next_
                        index += 1
                        current_column += 1
                    elif next_.isalpha() or next_ == '_':
                        return LexicalReturnStatus.INVALID_NUMBER_FORMAT, (current_line,current_column), token_list, symbol_table
                    else:
                        state = LexicalAnalyzerStates.WHITE_SPACE
                        token_list.append(Token(Terminal.INT_CONST, current_line, token_column, current_token))
                        current_token = ''
                case LexicalAnalyzerStates.FLOAT_FRACTIONAL:
                    if next_.isdigit():
                        current_token += next_
                        index += 1
                        current_column += 1
                    elif next_.isalpha() or next_ == '_':
                        return LexicalReturnStatus.INVALID_NUMBER_FORMAT, (current_line,current_column), token_list, symbol_table
                    else:
                        state = LexicalAnalyzerStates.WHITE_SPACE
                        token_list.append(Token(Terminal.FLOAT_CONST, current_line, token_column, current_token))
                        current_token = ''
                case LexicalAnalyzerStates.STRING_LITERAL:
                    if next_ == '"':
                        state = LexicalAnalyzerStates.WHITE_SPACE
                        token_list.append(Token(Terminal.STRING_CONST, current_line, token_column, current_token))
                        current_token = ''
                    elif next_ == '\n':
                        return LexicalReturnStatus.UNCLOSED_STRING, (current_line,current_column), token_list, symbol_table
                    else:
                        current_token += next_

                    index += 1
                    current_column += 1
                case LexicalAnalyzerStates.ASSIGN_OR_EQUAL:
                    state = LexicalAnalyzerStates.WHITE_SPACE
                    if next_ == '=':
                        token_list.append(Token(Terminal.EQUAL, current_line, token_column))
                        index += 1
                        current_column += 1
                    else:
                        token_list.append(Token(Terminal.ASSIGNMENT, current_line, token_column))
                    current_token = ''
                case LexicalAnalyzerStates.GREATER_OR_GE:
                    state = LexicalAnalyzerStates.WHITE_SPACE
                    if next_ == '=':
                        token_list.append(Token(Terminal.GREATER_EQUAL, current_line, token_column))
                        index += 1
                        current_column += 1
                    else:
                        token_list.append(Token(Terminal.GREATER_THAN, current_line, token_column))
                    current_token = ''
                case LexicalAnalyzerStates.LESS_OR_LE:
                    state = LexicalAnalyzerStates.WHITE_SPACE
                    if next_ == '=':
                        token_list.append(Token(Terminal.LESS_EQUAL, current_line, token_column))
                        index += 1
                        current_column += 1
                    else:
                        token_list.append(Token(Terminal.LESS_THAN, current_line, token_column))
                    current_token = ''
                case LexicalAnalyzerStates.DIFFERENT:
                    state = LexicalAnalyzerStates.WHITE_SPACE
                    if next_ == '=':
                        token_list.append(Token(Terminal.NOT_EQUAL, current_line, token_column))
                        index += 1
                        current_column += 1
                    else:
                        return LexicalReturnStatus.INVALID_CHAR, (current_line,current_column), token_list, symbol_table
                    current_token = ''
                case _:
                    print('\n\nWhat\n\n')

        return LexicalReturnStatus.OK, (0,0), token_list, symbol_table
